### PR TITLE
[PREFERENCE-SERVICE] Create Preference Entity, Service, Controller an…

### PR DESCRIPTION
**Pull request by aacgn**
Solves #23

### CODE REVIEW

File: preference-service/src/controllers/preference.controller.ts [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+        if (!userId || !placeCategoryId || !checkoutSlotId) {
+            res.status(HttpStatus.BAD_REQUEST).json(new ServerError(
+                Messages.validation.preference_data_needs_to_be_provided,
+                ErrorCode.NOT_ENOUGH_DATA,
+            ).toJSON());
+
+            return;
+        }
+
+        if (isNaN(userId)) {
```
Comment (gilesv) [inline preference-service/src/controllers/preference.controller.ts]: [userId, placeCategoryId, checkoutSlotId].forEach(param => {
            if (isNaN(param)) {
                res.status(HttpStatus.BAD_REQUEST).json(new ServerError(
                    Messages.validation.id_must_be_number(param),
                    ErrorCode.NOT_ENOUGH_DATA,
                ).toJSON());
            }
        });

Comment (viniciusgiles) [inline preference-service/src/controllers/preference.controller.ts]:
```typescript
[userId, placeCategoryId, checkoutSlotId].forEach(param => {
    if (isNaN(param)) {
        res.status(HttpStatus.BAD_REQUEST).json(new ServerError(
            Messages.validation.id_must_be_number(param),
            ErrorCode.NOT_ENOUGH_DATA,
        ).toJSON());
    }
});
```

Comment (aacgn) [reply preference-service/src/controllers/preference.controller.ts]: Done

File: preference-service/src/entities/preference.entity.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
+import {
+    Column,
+    Entity,
+    PrimaryGeneratedColumn,
+    CreateDateColumn,
+    UpdateDateColumn,
+    ManyToOne,
```
Comment (eduardosm7) [inline preference-service/src/entities/preference.entity.ts]: Unused imports

File: preference-service/src/services/preference.service.ts [hunk 3]
```diff
@@ -0,0 +1,88 @@
+import db from "../config/database";
+import { Preference } from "../entities/preference.entity";
+import { ServerError, ErrorCode } from "../utils/serverError";
+import { ValidationError, validate } from "class-validator";
+import Messages from "../utils/Messages";
```
Comment (eduardosm7) [inline preference-service/src/services/preference.service.ts]: Unused import

File: preference-service/src/services/preference.service.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * @namespace Services
+ * @class PreferenceService
+ */
+export class PreferenceService {
+    /**
+     * Create order in database
```
Comment (eduardosm7) [inline preference-service/src/services/preference.service.ts]: Wrong comment

File: preference-service/src/controllers/preference.controller.ts [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        const id = req.params.id;
+
+        if (isNaN(id)) {
+            res.status(HttpStatus.BAD_REQUEST).json(new ServerError(
+                Messages.validation.id_must_be_number("id"),
+                ErrorCode.NOT_ENOUGH_DATA,
+            ).toJSON());
+        }
+
+        const permissions = await PreferenceService.findById(id);
```
Comment (eduardosm7) [inline preference-service/src/controllers/preference.controller.ts]: wrong constant naming

File: preference-service/src/controllers/preference.controller.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        const userId = req.params.userId;
+
+        if (isNaN(userId)) {
+            res.status(HttpStatus.BAD_REQUEST).json(new ServerError(
+                Messages.validation.id_must_be_number("userId"),
+                ErrorCode.NOT_ENOUGH_DATA,
+            ).toJSON());
+        }
+
+        const permissions = await PreferenceService.findByUserId(userId);
```
Comment (eduardosm7) [inline preference-service/src/controllers/preference.controller.ts]: wrong constant naming

File: preference-service/src/controllers/preference.controller.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+        const placeCategoryId = req.params.placeCategoryId;
+
+        if (isNaN(placeCategoryId)) {
+            res.status(HttpStatus.BAD_REQUEST).json(new ServerError(
+                Messages.validation.id_must_be_number("placeCategoryId"),
+                ErrorCode.NOT_ENOUGH_DATA,
+            ).toJSON());
+        }
+
+        const permissions = await PreferenceService.findByPlaceCategoryId(placeCategoryId);
```
Comment (eduardosm7) [inline preference-service/src/controllers/preference.controller.ts]: wrong constant naming

File: preference-service/src/controllers/preference.controller.ts [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+        res.json(permissions);
+    }
+
+    public static async getAllPreferences(req: Request, res: Response) {
+        const permissions = await PreferenceService.findAll();
```
Comment (eduardosm7) [inline preference-service/src/controllers/preference.controller.ts]: wrong constant naming

File: preference-service/src/routes/preference.route.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+
+
+function PreferenceRouter(): Router {
+    const router = Router();
+
+    router.post("/create", PreferenceController.createPreference);
+    router.get("/", PreferenceController.getAllPreferences);
+    router.get("/:id", PreferenceController.getPreferenceById);
+    router.get("/user/:userId", PreferenceController.getPreferenceByUserId);
+    router.get("/place-categories/:placeCategoryId", PreferenceController.getPreferenceByPlaceCategoryId);
```
Comment (eduardosm7) [inline preference-service/src/routes/preference.route.ts:12]: Following the pattern, I believe it would be place-category

**Review by eduardosm7 — CHANGES_REQUESTED** [review]

File: preference-service/src/routes/preference.route.ts (hunk 9 above)
Comment (viniciusgiles) [reply preference-service/src/routes/preference.route.ts:12]: All endpoints in core-api are in plural form... I suppose it's better to change `/user/` to `/users/`

Comment (eduardosm7) [reply preference-service/src/routes/preference.route.ts:12]: Done

File: preference-service/src/entities/preference.entity.ts (hunk 2 above)
Comment (eduardosm7) [reply preference-service/src/entities/preference.entity.ts]: Done

File: preference-service/src/services/preference.service.ts (hunk 4 above)
Comment (eduardosm7) [reply preference-service/src/services/preference.service.ts]: Done

File: preference-service/src/services/preference.service.ts (hunk 3 above)
Comment (eduardosm7) [reply preference-service/src/services/preference.service.ts]: Done

File: preference-service/src/controllers/preference.controller.ts (hunk 5 above)
Comment (eduardosm7) [reply preference-service/src/controllers/preference.controller.ts]: Done

File: preference-service/src/controllers/preference.controller.ts (hunk 6 above)
Comment (eduardosm7) [reply preference-service/src/controllers/preference.controller.ts]: Done

File: preference-service/src/controllers/preference.controller.ts (hunk 7 above)
Comment (eduardosm7) [reply preference-service/src/controllers/preference.controller.ts]: Done

File: preference-service/src/controllers/preference.controller.ts (hunk 8 above)
Comment (eduardosm7) [reply preference-service/src/controllers/preference.controller.ts]: Done

File: preference-service/src/routes/preference.route.ts (hunk 9 above)
Comment (eduardosm7) [reply preference-service/src/routes/preference.route.ts:12]: Done

File: preference-service/src/controllers/preference.controller.ts (hunk 1 above)
Comment (viniciusgiles) [reply preference-service/src/controllers/preference.controller.ts]: @aacgn Now I see a problem with my approach. This is better:
```typescript
for (let param of [userId, placeCategoryId, checkoutSlotId])
    if (isNaN(param)) {
        res.status(HttpStatus.BAD_REQUEST).json(new ServerError(
            Messages.validation.id_must_be_number(param),
            ErrorCode.NOT_ENOUGH_DATA,
        ).toJSON());
        return;
    }
};
```

**Review by eduardosm7 — APPROVED** [review]

**Pull request merged**